### ISSUE TITLE
[pimatic-shell-execute] API is unable to toggle ShellSwitch

### ISSUE THREAD

**Issue by MaxWinterstein** [pimatic/pimatic#783]
Hello,

i have a ShellSwitch Device, copied from https://github.com/pimatic/pimatic-shell-execute#shellswitch-device . 

When i try to toggle it via api 
```
/api/device/light/toggle
```
it is only possible to turn the device on, not off. 

I think toggle is looking at the current state via the "getState" command, even this is not used in the device?

**Comment by Icesory** [comment]
At the moment "getState"  runs the code defined in the "getStateCommand" field. And when you set it up with "echo false" it returns ever false and toggle can only switch to true. 
https://github.com/pimatic/pimatic-shell-execute/blob/master/shell-execute.coffee

The getState command should check his config and when "getStateCommand" isn't set it should return the state of the device in pimatic.

**Comment by MaxWinterstein** [comment]
> The getState command should check his config and when "getStateCommand" isn't set it should return the state of the device in pimatic.

Yes, exactly that would i expect. I have no "getStateCommand" and i think it fallsback to default. 
https://github.com/pimatic/pimatic-shell-execute/blob/master/device-config-schema.coffee#L21

This should be implemented, it is not possible to get a state out of every thing ;)

**Comment by mwittig** [comment]
@Icesory I think, it is better to remove remove the default from the schema as it does not make sense really. Some users may think it is enough to give the on and off commands, but it is not if you want to have consistent state in pimatic. As it may some sometimes difficult to get the state of the external device, may be we should implement a state variable which is set by default if the on or off command is executed.  For this, I would set the 'getStateCommand' as being optional.

**Comment by Icesory** [comment]
I have looked at the code and an simple remove of the default state will not help.
But i will create an fix now.

**Comment by sweetpi** [comment]
> @Icesory I think, it is better to remove remove the default from the schema as it does not make sense really. Some users may think it is enough to give the on and off commands, but it is not if you want to have consistent state in pimatic. As it may some sometimes difficult to get the state of the external device, may be we should implement a state variable which is set by default if the on or off command is executed. For this, I would set the 'getStateCommand' as being optional.



> I have looked at the code and an simple remove of the default state will not help.
> But i will create an fix now.

Sounds like a good solution :+1:

**Comment by mwittig** [comment]
@icesory Did you already work on a fix? If not I'll be happy to look into this.

**Comment by Icesory** [comment]
Sorry. I have forgot this issue.
No i have not work on an fix for this. But this should be only a one liner to fix.

```coffeescript
    getState: () ->
      if !@config.getStateCommand? then return Promise.resolve @_state
      return exec(@config.getStateCommand).then( (streams) =>
        stdout = streams[0]
        stderr = streams[1]
        stdout = stdout.trim()
        
        switch stdout
          when "on", "true", "1", "t", "o"
            @_setState(on)
            @_state = on
            return Promise.resolve @_state
          when "off", "false", "0", "f"
            @_setState(off)
            @_state = off
            return Promise.resolve @_state
          else
            env.logger.error "ShellSwitch: stderr output from getStateCommand for #{@name}: #{stderr}" if stderr.length isnt 0
            throw new Error "ShellSwitch: unknown state=\"#{stdout}\"!"
        )
```
and of course the default for the getCommand must be removed

**Comment by mwittig** [comment]
Ok. Implemented it as part of PR https://github.com/pimatic/pimatic-shell-execute/pull/8

**Comment by mwittig** [comment]
@MaxWinterstein Fixed in release 0.8.14.
